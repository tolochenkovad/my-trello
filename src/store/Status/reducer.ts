import { createReducer, AsyncThunk, AnyAction } from '@reduxjs/toolkit';
import { toastr } from 'react-redux-toastr';

interface IBaseStatusPayload {
  message?: string;
  id?: string | number;
  statusCode?: number;
}

export interface StatusPayloadFromBackend extends IBaseStatusPayload {
  error?: any;
}

export interface StatusPayload extends IBaseStatusPayload {
  error?: any;
}

type GenericAsyncThunk = AsyncThunk<unknown, StatusPayloadFromBackend, any>;

type PendingAction = ReturnType<GenericAsyncThunk['pending']>;
type RejectedAction = ReturnType<GenericAsyncThunk['rejected']>;
type FulfilledAction = ReturnType<GenericAsyncThunk['fulfilled']>;

function isPendingAction(action: AnyAction): action is PendingAction {
  return !!action.type.endsWith('/pending');
}

function isRejectedAction(action: AnyAction): action is RejectedAction {
  return !!action.type.endsWith('/rejected');
}

function isFulfilledAction(action: AnyAction): action is FulfilledAction {
  return !!action.type.endsWith('/fulfilled');
}

type StatusTypes = 'LOADING' | 'ERROR';

interface IStatusInitialState {
  [action: string]: {
    status: StatusTypes;
  } & StatusPayload;
}

const initialState: IStatusInitialState = {};

const getActionType = (type: string) => type.split('/').slice(0, -1).join('/');

const statusReducer = createReducer(initialState, (builder) => {
  builder
    .addMatcher(isPendingAction, (state, action) => {
      const actionType = getActionType(action.type);

      state[actionType] = {
        status: 'LOADING',
      };
    })
    .addMatcher(isRejectedAction, (state, action) => {
      const actionType = getActionType(action.type);

      state[actionType] = {
        status: 'ERROR',
        error: action.error,
        message: 'ERROR!',
      };

      toastr.error('', 'ERROR!');
    })
    .addMatcher(isFulfilledAction, (state, action) => {
      const actionType = getActionType(action.type);
      delete state[actionType];
    });
});

export default statusReducer;
