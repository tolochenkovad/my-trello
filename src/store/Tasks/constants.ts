export const SET_TASKS = `APP/tasks/SET_TASKS` as const;
export const REMOVE_TASK = `APP/tasks/REMOVE_TASK` as const;

export const SAVE_DATA = `APP/tasks/SAVE_DATA` as const;

export const SET_ERROR = `APP/tasks/SET_ERROR` as const;
export const SET_LOADING = `APP/tasks/SET_LOADING` as const;
