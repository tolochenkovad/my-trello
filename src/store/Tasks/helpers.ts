import { last, size } from 'lodash';
import { initialDataType } from '../../types/tasks';

function sortTasksIds(a: string, b: string): number {
  if (a.includes('-') && b.includes('-')) {
    const leftElement = last(a.split('-'));
    const rightElement = last(b.split('-'));
    if (leftElement && rightElement) {
      return +leftElement - +rightElement;
    }
  }
  return -1;
}

export function getIndexForNewTask(tasks: initialDataType['tasks']): number {
  const lastKey = size(tasks) > 0 && last(Object.keys(tasks).sort(sortTasksIds));
  const lastIndex = lastKey ? last(lastKey.split('-')) : 0;
  return lastIndex ? +lastIndex + 1 : 1;
}
