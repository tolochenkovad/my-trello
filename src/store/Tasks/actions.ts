import * as Constants from './constants';
import { initialDataType } from '../../types/tasks';
import { toastr } from 'react-redux-toastr';
import { ThunkResult } from '../../types';
import { addDoc, getCollectionsFromFirebase, removeData, updateData } from '../../api/firebase/api';
import { COLLECTIONS } from '../../constants';
import { getIndexForNewTask } from './helpers';
import { size } from 'lodash';
import moment from 'moment';

// thunks
export const getTasks = (): ThunkResult => async (dispatch, getState) => {
  dispatch(setLoading());
  try {
    const authId = getState().firebase.auth.uid;
    const data = await getCollectionsFromFirebase(COLLECTIONS.tasks, authId);
    dispatch(setTasks(data));
  } catch (error) {
    toastr.error(error, '');
    dispatch(setError(error));
  }
};

export const addTask = (value: string, color: string, dateOfTheEnd: string): ThunkResult => async (dispatch, getState) => {
  dispatch(setLoading());
  try {
    const authId = getState().firebase.auth.uid;
    const { tasks } = getState().tasks.dataForDraggable;
    const index = getIndexForNewTask(tasks);
    const newTask = {
      [`task-${index}`]: { id: `task-${index}`, content: value, columnId: 'column-1', date: moment().format(), color, dateOfTheEnd },
    };
    if (size(tasks) === 0) {
      await addDoc(COLLECTIONS.tasks, newTask, authId);
    } else {
      await updateData(COLLECTIONS.tasks, newTask, authId);
    }
    dispatch(getTasks());
  } catch (error) {
    toastr.error(error, '');
    dispatch(setError(error));
  }
};

export const editTask = (value: string, color: string, taskId: string, dateOfTheEnd: string): ThunkResult => async (dispatch, getState) => {
  dispatch(setLoading());
  try {
    const authId = getState().firebase.auth.uid;
    const { tasks } = getState().tasks.dataForDraggable;
    const updatedTask = { [taskId]: tasks[taskId] };
    updatedTask[taskId].content = value;
    updatedTask[taskId].date = moment().format();
    updatedTask[taskId].color = color;
    updatedTask[taskId].dateOfTheEnd = dateOfTheEnd;
    await updateData(COLLECTIONS.tasks, updatedTask, authId);
    dispatch(getTasks());
  } catch (error) {
    toastr.error(error, '');
    dispatch(setError(error));
  }
};

export const saveDataToServer = (data: initialDataType): ThunkResult => async (dispatch, getState) => {
  dispatch(setLoading());
  try {
    dispatch(saveData(data));
    const authId = getState().firebase.auth.uid;
    await updateData(COLLECTIONS.tasks, data.tasks, authId);
    dispatch(getTasks());
  } catch (error) {
    toastr.error(error, '');
    dispatch(setError(error));
  }
};

export const removeTask = (taskId: string): ThunkResult => async (dispatch, getState) => {
  dispatch(setLoading());
  try {
    const authId = getState().firebase.auth.uid;
    dispatch(deleteTask(taskId));
    await removeData(COLLECTIONS.tasks, taskId, authId);
    dispatch(getTasks());
  } catch (error) {
    toastr.error(error, '');
    dispatch(setError(error));
  }
};

// actions
export const setTasks = (tasks: initialDataType['tasks']) => ({
  type: Constants.SET_TASKS,
  payload: { tasks },
});

export const deleteTask = (taskId: string) => ({
  type: Constants.REMOVE_TASK,
  payload: { taskId },
});

const setError = (error: Error) => ({
  type: Constants.SET_ERROR,
  payload: { error },
});

const saveData = (data: initialDataType) => ({
  type: Constants.SAVE_DATA,
  payload: { data },
});

export const setLoading = () => ({
  type: Constants.SET_LOADING,
});

export type Actions =
  | ReturnType<typeof setTasks>
  | ReturnType<typeof deleteTask>
  | ReturnType<typeof saveData>
  | ReturnType<typeof setError>
  | ReturnType<typeof setLoading>;
