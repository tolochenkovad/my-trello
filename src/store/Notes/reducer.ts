import * as Constants from './constants';
import { Actions } from './actions';

export const initialState = {
  note: 'test' as string,
  isReady: false,
  data: { isTest: false },
};

const notesReducer = (state = initialState, action: Actions): typeof initialState => {
  switch (action.type) {
    case Constants.GET_NOTES:
      return {
        ...state,
        isReady: action.payload.test,
      };
    case Constants.GET_CATEGORIES:
      return {
        ...state,
        isReady: action.payload.isReady,
      };
    default:
      return state;
  }
};

export default notesReducer;