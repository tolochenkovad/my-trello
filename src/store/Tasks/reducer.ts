import { forEach } from 'lodash';
import * as Constants from './constants';
import { Actions } from './actions';
import { initialDataType } from '../../types/tasks';

export const INITIAL_DATA = {
  tasks: {},
  columns: {
    'column-1': {
      id: 'column-1',
      title: 'To do',
      taskIds: [],
    },
    'column-2': {
      id: 'column-2',
      title: 'In progress',
      taskIds: [],
    },
    'column-3': {
      id: 'column-3',
      title: 'Done',
      taskIds: [],
    },
  },
  columnOrder: ['column-1', 'column-2', 'column-3'],
};

export type initialStateTypesForReducer = {
  dataForDraggable: initialDataType;
  isLoading: boolean;
  error: string;
};

export const initialState: initialStateTypesForReducer = {
  dataForDraggable: INITIAL_DATA,
  isLoading: false as boolean,
  error: '',
};

const tasksReducer = (state = initialState, action: Actions): initialStateTypesForReducer => {
  switch (action.type) {
    case Constants.SET_LOADING:
      return {
        ...state,
        isLoading: true,
      };
    case Constants.SET_TASKS: {
      const { tasks } = action.payload;
      const copyDataForDraggable: initialDataType = JSON.parse(JSON.stringify(state.dataForDraggable));
      forEach(copyDataForDraggable.columns, (columnItem) => {
        forEach(tasks, (taskItem) => {
          if (taskItem.columnId === columnItem.id) {
            columnItem.taskIds.push(taskItem.id);
          }
        });
      });
      copyDataForDraggable.tasks = tasks;

      return {
        ...state,
        dataForDraggable: copyDataForDraggable,
        isLoading: false,
      };
    }

    case Constants.SET_ERROR:
      return {
        ...state,
        isLoading: false,
        error: action.payload.error.message,
      };
    default:
      return state;
  }
};

export default tasksReducer;
