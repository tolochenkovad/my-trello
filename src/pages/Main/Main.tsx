import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { DragDropContext, DropResult } from 'react-beautiful-dnd';
import { initialDataType } from '../../types/tasks';
import { INITIAL_DATA } from '../../store/Tasks/reducer';
import { getTasks } from '../../store/Tasks/actions';
import Column from '../../components/Column';
import { AppStore } from '../../types';
import { useDispatch, useSelector } from 'react-redux';

const Main: React.FC = () => {
  const [state, setState] = useState<initialDataType>(INITIAL_DATA);
  const [value, setValue] = useState('');
  const dispatch = useDispatch();
  const tasks = useSelector((state: AppStore) => state.tasks.dataForDraggable);

  useEffect(() => {
    if (tasks.columnOrder.length > 0) {
      setState(tasks);
    } else {
      dispatch(getTasks());
    }
  }, [dispatch, tasks]);

  const onDragEnd = useCallback(
    (result: DropResult) => {
      const { destination, source, draggableId } = result;

      if (!destination) {
        return;
      }

      if (destination.droppableId === source.droppableId && destination.index === source.index) {
        return;
      }

      const start = state.columns[source.droppableId];
      const finish = state.columns[destination.droppableId];

      if (start === finish) {
        const newTaskIds = Array.from(start.taskIds);
        newTaskIds.splice(source.index, 1);
        newTaskIds.splice(destination.index, 0, draggableId);

        const newColumn = {
          ...start,
          taskIds: newTaskIds,
        };

        const newState = {
          ...state,
          columns: {
            ...state.columns,
            [newColumn.id]: newColumn,
          },
        };

        setState(newState);
        return;
      }

      // Moving from one list to another
      const startTaskIds = Array.from(start.taskIds);
      startTaskIds.splice(source.index, 1);
      const newStart = {
        ...start,
        taskIds: startTaskIds,
      };

      const finishTaskIds = Array.from(finish.taskIds);
      finishTaskIds.splice(destination.index, 0, draggableId);
      const newFinish = {
        ...finish,
        taskIds: finishTaskIds,
      };

      const newState = {
        ...state,
        columns: {
          ...state.columns,
          [newStart.id]: newStart,
          [newFinish.id]: newFinish,
        },
      };
      setState(newState);
    },
    [state],
  );

  const renderDragContainer = useMemo(
    () => (
      <DragDropContext onDragEnd={onDragEnd}>
        <div className="container">
          {state.columnOrder.map((columnId) => {
            const column = state.columns[columnId];
            const tasks = column.taskIds.map((taskId) => state.tasks[taskId]);

            return <Column key={column.id} column={column} tasks={tasks} />;
          })}
        </div>
      </DragDropContext>
    ),
    [onDragEnd, state],
  );

  console.log('render');
  return (
    <>
      {renderDragContainer}
      <input type="text" value={value} onChange={(e) => setValue(e.target.value)} />
    </>
  );
};

export default Main;
