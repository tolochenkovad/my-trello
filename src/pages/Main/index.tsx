import React, { FC, useCallback, useEffect, useRef, useState } from 'react';
import { DragDropContext, DropResult } from 'react-beautiful-dnd';
import { initialDataType } from '../../types/tasks';
import { INITIAL_DATA } from '../../store/Tasks/reducer';
import Column from '../../components/Column';
import { useDispatch, useSelector } from 'react-redux';
import { getDataForDraggable } from '../../store/Tasks/selectors';
import { getTasksAction, saveDataToServerAction } from '../../store/Tasks/actions';

const Main: FC = () => {
  const [state, setState] = useState<initialDataType>(INITIAL_DATA);

  const dispatch = useDispatch();
  const dataForDraggable = useSelector(getDataForDraggable);
  const isMounted = useRef(false);

  useEffect(() => {
    if (!isMounted.current) {
      dispatch(getTasksAction.pending({}));
      isMounted.current = true;
    } else {
      setState(dataForDraggable);
    }
  }, [dispatch, dataForDraggable]);

  const onDragEnd = useCallback(
    (result: DropResult) => {
      const { destination, source, draggableId } = result;

      if (!destination) {
        return;
      }

      if (destination.droppableId === source.droppableId && destination.index === source.index) {
        return;
      }

      const start = state.columns[source.droppableId];
      const finish = state.columns[destination.droppableId];

      if (start === finish) {
        const newTaskIds = Array.from(start.taskIds);
        newTaskIds.splice(source.index, 1);
        newTaskIds.splice(destination.index, 0, draggableId);

        const newColumn = {
          ...start,
          taskIds: newTaskIds,
        };

        const newState = {
          ...state,
          columns: {
            ...state.columns,
            [newColumn.id]: newColumn,
          },
        };

        setState(newState);
        return;
      }

      // Moving from one list to another
      const startTaskIds = Array.from(start.taskIds);
      startTaskIds.splice(source.index, 1);
      const newStart = {
        ...start,
        taskIds: startTaskIds,
      };

      const finishTaskIds = Array.from(finish.taskIds);
      finishTaskIds.splice(destination.index, 0, draggableId);
      const newFinish = {
        ...finish,
        taskIds: finishTaskIds,
      };

      const newState = {
        ...state,
        columns: {
          ...state.columns,
          [newStart.id]: newStart,
          [newFinish.id]: newFinish,
        },
        tasks: {
          ...state.tasks,
          [draggableId]: {
            ...state.tasks[draggableId],
            columnId: newFinish.id,
          },
        },
      };
      setState(newState);

      dispatch(saveDataToServerAction.pending(newState));
    },
    [dispatch, state],
  );

  return (
    <DragDropContext onDragEnd={onDragEnd}>
      <div className="d-flex">
        {state.columnOrder.map((columnId) => {
          const column = state.columns[columnId];
          const tasks = column.taskIds.map((taskId) => state.tasks[taskId]);
          return <Column key={column.id} column={column} tasks={tasks} />;
        })}
      </div>
    </DragDropContext>
  );
};

export default Main;
