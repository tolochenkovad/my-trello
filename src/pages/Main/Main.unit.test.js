import React from 'react';
import { mount, shallow } from 'enzyme';
import { size } from 'lodash';
import Main from './Main';
import { useSelector, useDispatch } from 'react-redux';
import { INITIAL_DATA } from '../../store/Tasks/reducer';


let wrapper;

useSelector.mockImplementation(() => INITIAL_DATA);

describe('Main', () => {
  describe('shallow', () => {
    beforeEach(() => {
      wrapper = shallow(<Main />);
    });
    it('should render', () => {
      expect(wrapper).toHaveLength(1);
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe('children is exist', () => {
      it('DragDropContext', () => {
        expect(wrapper.find('DragDropContext')).toHaveLength(1);
      });

      it('Column', () => {
        expect(wrapper.find('Memo(Column)')).toHaveLength(size(INITIAL_DATA.columnOrder));
      });
    });
  });

  describe('should works correctly', () => {
    it('should call getTasks if component is mounted', () => {
      const getTasks = jest.fn();
      useDispatch.mockImplementation(() => getTasks);
      wrapper = mount(<Main />);
      expect(getTasks).toHaveBeenCalled();
    });
  });
});
