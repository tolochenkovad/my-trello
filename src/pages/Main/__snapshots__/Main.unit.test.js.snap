// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main shallow should render correctly 1`] = `
<DragDropContext
  onDragEnd={[Function]}
>
  <div
    className="container"
  >
    <Memo(Column)
      column={
        Object {
          "id": "column-1",
          "taskIds": Array [],
          "title": "To do",
        }
      }
      key="column-1"
      tasks={Array []}
    />
    <Memo(Column)
      column={
        Object {
          "id": "column-2",
          "taskIds": Array [],
          "title": "In progress",
        }
      }
      key="column-2"
      tasks={Array []}
    />
    <Memo(Column)
      column={
        Object {
          "id": "column-3",
          "taskIds": Array [],
          "title": "Done",
        }
      }
      key="column-3"
      tasks={Array []}
    />
  </div>
</DragDropContext>
`;

exports[`Main should render correctly 1`] = `
<DragDropContext
  onDragEnd={[Function]}
>
  <div
    className="container"
  >
    <Memo(Column)
      column={
        Object {
          "id": "column-1",
          "taskIds": Array [],
          "title": "To do",
        }
      }
      key="column-1"
      tasks={Array []}
    />
    <Memo(Column)
      column={
        Object {
          "id": "column-2",
          "taskIds": Array [],
          "title": "In progress",
        }
      }
      key="column-2"
      tasks={Array []}
    />
    <Memo(Column)
      column={
        Object {
          "id": "column-3",
          "taskIds": Array [],
          "title": "Done",
        }
      }
      key="column-3"
      tasks={Array []}
    />
  </div>
</DragDropContext>
`;
